### PR TITLE
42 redo sbn reintegrate

### PR DESCRIPTION
**Pull request by williamzhang0306**
This branch updates the sbn_adapter data source to the current state of OnAir and including unit tests.

### CODE REVIEW

**Comment by williamzhang0306** [comment]
Should the files config/sbn_cfs_config and data.telemetry_configs/adapter_TLM_CONFIG.json stay in this branch? I'm pretty sure they were used for testing sbn_adapter with cfs at one point, but I don't know where that testing was done.

File: onair/data_handling/sbn_adapter.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    def get_next(self):
+        """Provides the latest data from SBN in a dictionary of lists structure.
+        Returned data is safe to use until the next get_next call.
+        Blocks until new data is available."""
+
+        data_available = False
+
+        while not data_available:
+            with self.new_data_lock:
+                data_available = self.has_data()
```
Comment (the-other-james) [inline onair/data_handling/sbn_adapter.py:128]: Why is the has_data() function needed here instead of just checking self.new_data? Is this needed to ease unit testing?

File: onair/data_handling/sbn_adapter.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            # prints out the data from the message to the terminal
+            print(", ".join([field_name + ": " + str(getattr(recv_msg, field_name)) for field_name, field_type in recv_msg._fields_[1:]]))
+
+            # TODO: Lock needed here?
+            self.get_current_data(recv_msg, data_struct, app_name)
+
+    def get_current_data(self, recv_msg, data_struct, app_name):
+        # TODO: Lock needed here?
+        current_buffer = self.currentData[(self.double_buffer_read_index + 1) %2]
+        secondary_header = recv_msg.TlmHeader.Secondary
```
Comment (the-other-james) [inline onair/data_handling/sbn_adapter.py:170]: Why did you add back in the timing field?

File: onair/data_handling/sbn_adapter.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                for sub_type in name.split('.'):
+                    current_object = getattr(current_object, sub_type)
+                    data = str(current_object) # note does not work for arrays?
+                current_buffer['data'][idx] = data
+
+        with self.new_data_lock:
+            self.new_data = True
+
+    def has_data(self):
+        return self.new_data
```
Comment (the-other-james) [inline onair/data_handling/sbn_adapter.py:198]: Please leave newlines at the end of files (it makes it look nicer when someone cats a file in their terminal)

File: test/onair/data_handling/test_sbn_adapter.py [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+    # field type was not defined in message_headers.py and has no subfields of its own
+    field_type.__str__ = MagicMock()
+    field_type.__str__.return_value = 'fooble'
+    del field_type._fields_
+
+    # Act
+    result = cut.gather_field_names(field_name, field_type)
+
+    # Assert
+    # TODO return type depends on bugfix(?) in
```
Comment (the-other-james) [inline test/onair/data_handling/test_sbn_adapter.py:112]: Can this TODO be removed?

**Review by the-other-james — CHANGES_REQUESTED** [review]
Overall this is looking really good! I had just a few questions / nitpicks.

Can you also document how you ran this with cFS? If this is short, you can add it to the top level readme, otherwise do it in a separate doc. For cFS, can you point me to the repo you set up and I'll fork it?

Great work, thanks!

**Comment by the-other-james** [comment]
I copied these changes over to a setup I have for a different project, and can confirm that it functions as expected and the OnAIR results were identical with the older SBN rework I did.

File: onair/data_handling/sbn_adapter.py (hunk 2 above)
Comment (the-other-james) [reply onair/data_handling/sbn_adapter.py:170]: Oh, it looks like I removed that in a different version. Apologies!

We don't want to force any assumptions about time, so these lines should all be removed:

```
         secondary_header = recv_msg.TlmHeader.Secondary
         #gets seconds from header and adds to current buffer
         start_time = datetime.datetime(1969, 12, 31, 20) # January 1, 1980
         seconds = secondary_header.Seconds
         subseconds = secondary_header.Subseconds
         curr_time = seconds + (2**(-32) * subseconds) # a subsecond is equal to 2^-32 second
         time = start_time + datetime.timedelta(seconds=curr_time)
         str_time = time.strftime("%Y-%j-%H:%M:%S.%f")
         current_buffer['data'][0] = str_time
```

File: onair/data_handling/sbn_adapter.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+        # Launch thread to listen for messages
+        self.listener_thread = threading.Thread(target=self.message_listener_thread)
+        self.listener_thread.start()
+
+        # subscribe to message IDs
+        for msgID in self.msgID_lookup_table.keys():
+            sbn.subscribe(msgID)
+
+    # TODO!!!!! Bug? Where levels of nesting affect return type
```
Comment (the-other-james) [inline onair/data_handling/sbn_adapter.py:58]: I think this TODO can be removed now, no?

File: onair/data_handling/sbn_adapter.py (hunk 1 above)
Comment (williamzhang0306) [reply onair/data_handling/sbn_adapter.py:128]: Yes this change was made to have more control during unit testing. Additionally, I saw that redis_adapter had also used a has_data() function instead of just checking self.new_data.

**Comment by williamzhang0306** [comment]
> Overall this is looking really good! I had just a few questions / nitpicks.
> 
> Can you also document how you ran this with cFS? If this is short, you can add it to the top level readme, otherwise do it in a separate doc. For cFS, can you point me to the repo you set up and I'll fork it?
> 
> Great work, thanks!

Thank you! This is the repo I used for cfs: https://github.com/williamzhang0306/cFS-OnAIR.

**Comment by jlucas9** [comment]
Note that we are also including OnAir in NOS3 which includes a distribution of cFS draco-rc4 at the time of writing:
* https://github.com/nasa/nos3/tree/nos3%23298-onair-integration

We made forks for now trying to get things working:
* https://github.com/nasa-itc/OnAIR
* https://github.com/nasa-itc/SBN-Client
* https://github.com/nasa-itc/SBN

I can add you both to the repos if you'd like to assist directly there and collaborate more closely!

File: onair/data_handling/sbn_adapter.py (hunk 2 above)
Comment (the-other-james) [reply onair/data_handling/sbn_adapter.py]: Thanks for removing!

File: onair/data/telemetry_configs/adapter_TLM_CONFIG.json [hunk 6]
```diff
@@ -115,5 +115,10 @@
...
     "SAMPLE.sample_data_gps_t.sample_data_lng",
     "SAMPLE.sample_data_gps_t.sample_data_alt"
-  ]
+  ],
+  "channels":
+    {"0x0887": ["SAMPLE", "sample_data_power_t"],
+     "0x0889": ["SAMPLE", "sample_data_thermal_t"],
+     "0x088A": ["SAMPLE", "sample_data_gps_t"]
+    }
 }
```
Comment (the-other-james) [inline onair/data/telemetry_configs/adapter_TLM_CONFIG.json:124]: Forgot newline

**Review by the-other-james — CHANGES_REQUESTED** [review]
This is great work! I tried out the instructions and they worked flawlessly on my machine. Thanks!

Just a few small tweaks, and I want @asgibson to have a chance to review the unit tests before final approval.

**Review by asgibson — APPROVED** [review]
Unit tests look good enough.

**Review by the-other-james — APPROVED** [review]
